java/src/test/java/ai/onnxruntime/OnnxTensorTest.java

```diff
@@ -87,6 +87,83 @@ public void testScalarCreation() throws OrtException {
     }
   }
 
+  @Test
+  public void testBufferCreation() throws OrtException {
+    OrtEnvironment env = OrtEnvironment.getEnvironment();
+
+    float[] arrValues = new float[] {0, 1, 2, 3, 4};
+    try (OnnxTensor t = OnnxTensor.createTensor(env, arrValues)) {
+      // array creation isn't backed by buffers
+      Assertions.assertFalse(t.isCopy());
+      Assertions.assertFalse(t.getBufferRef().isPresent());
+      FloatBuffer buf = t.getFloatBuffer();
+      float[] output = new float[arrValues.length];
+      buf.get(output);
+      Assertions.assertArrayEquals(arrValues, output);
+
+      // Can't modify the tensor through this buffer.
+      buf.put(0, 25);
+      Assertions.assertArrayEquals(arrValues, output);
+    }
+
+    FloatBuffer nonDirectBuffer = FloatBuffer.allocate(5);
+    nonDirectBuffer.put(arrValues);
+    nonDirectBuffer.rewind();
+    try (OnnxTensor t = OnnxTensor.createTensor(env, nonDirectBuffer, new long[] {1, 5})) {
+      // non-direct buffers trigger a copy
+      Assertions.assertTrue(t.isCopy());
+      // tensors backed by buffers can get the buffer ref back out
+      Assertions.assertTrue(t.getBufferRef().isPresent());
+      FloatBuffer buf = t.getFloatBuffer();
+      float[] output = new float[arrValues.length];
+      buf.get(output);
+      Assertions.assertArrayEquals(arrValues, output);
+
+      // Can't modify the tensor through getFloatBuffer.
+      buf.put(0, 25);
+      Assertions.assertArrayEquals(arrValues, output);
+
+      // Can modify the tensor through getBufferRef.
+      FloatBuffer ref = (FloatBuffer) t.getBufferRef().get();
+      ref.put(0, 25);
+      buf = t.getFloatBuffer();
+      buf.get(output);
+      Assertions.assertEquals(25, output[0]);
+    }
+
+    FloatBuffer directBuffer =
+        ByteBuffer.allocateDirect(5 * 4).order(ByteOrder.nativeOrder()).asFloatBuffer();
+    directBuffer.put(arrValues);
+    directBuffer.rewind();
+    try (OnnxTensor t = OnnxTensor.createTensor(env, directBuffer, new long[] {1, 5})) {
+      // direct buffers don't trigger a copy
+      Assertions.assertFalse(t.isCopy());
+      // tensors backed by buffers can get the buffer ref back out
+      Assertions.assertTrue(t.getBufferRef().isPresent());
+      FloatBuffer buf = t.getFloatBuffer();
```

Comment:
The purpose of `getFloatBuffer()` methods is not clear. I would get rid of them. The purpose of the buffer access is to be direct without any overhead. If users want to make a copy, they can do so in their code and charge CPU usage and additional garbage produced to their code.
If you wanted to provide a read-only access, copying is an expensive way to do so when dealing with tensors.

Reply:
`getFloatBuffer` and the others are necessary to efficiently get the result out of an `OrtValue` allocated by ORT. Calling `getValue` to construct the multidimensional array is very slow as Java's multidimensional arrays are full of pointers. It's necessarily a copy as I can't expose the direct byte buffer which wraps the `OrtValue` pointer because I'll lose control of it's lifetime and the `OrtValue` could be deallocated while the user holds a reference to it. I could expose element-wise access but that would cross the JNI boundary for each access which is also slow.